src/main/java/com/google/sps/servlets/GetNearbyVendorsServlet.java

```diff
@@ -0,0 +1,142 @@
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    } catch (IllegalArgumentException e) {
+      System.out.println("Latitude and/or longitude outside legal range: " + e);
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+    // ---------------------------------------------------------------------------------
+    
+    // Check values exist and are in the range
+    if (prefixGeoHash.isEmpty() || distance > COMMONS.MAX_DISTANCE_CLIENT || distance < COMMONS.MIN_DISTANCE) {
+      System.out.println("The values do not exist and/or are outside the range.");
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+
+    Query query = buildGeoQuery(prefixGeoHash, hasDelivery);
+    
+    Iterable<Entity> vendorsRetrieved = fetchVendors(query);
+
+    List<Vendor> nearbyVendors = createVendorsList(vendorsRetrieved, clientLocation, distance);
+
+    Gson gson = new Gson();
+
+    response.setContentType("application/json;");
+    response.setCharacterEncoding("UTF-8");
+    response.getWriter().println(gson.toJson(nearbyVendors));
+  }
+
+  /** 
+   * Returns a query for a Vendor with a filter to match a substring (the prefix of a geoHash) and
+   * if the vendor has or not delivery service 
+   */
+  private Query buildGeoQuery(String prefixGeoHash, boolean hasDelivery) {
+    Filter lowerBound = new FilterPredicate("saleCard.location.geoHash",
+        FilterOperator.GREATER_THAN_OR_EQUAL, prefixGeoHash);
+    Filter upperBound = new FilterPredicate("saleCard.location.geoHash",
+        FilterOperator.LESS_THAN, prefixGeoHash + "\ufffd");
+    Filter deliveryFilter = new FilterPredicate("saleCard.hasDelivery",
+        FilterOperator.EQUAL, hasDelivery);
+
+    Filter geoFilter = CompositeFilterOperator.and(lowerBound, upperBound, deliveryFilter);
+
+    return new Query("Vendor").setFilter(geoFilter);
+  }
+
+  /** Returns an iterable with the results retrieved from Datastore */
+  private Iterable<Entity> fetchVendors(Query query) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    return datastore.prepare(query).asIterable();
+  }
+
+  /** Returns a list with the retrieved vendors and check they are within the requested distance*/
+  private List<Vendor> createVendorsList(Iterable<Entity> vendors, GeoPt clientLocation, float distanceLimit) {
+    List<Vendor> nearbyVendors = new ArrayList<>();
+    for (Entity vendorEntity : vendors) {
+      Vendor vendor = new Vendor(vendorEntity);
+      GeoPt vendorLocation = vendor.getBusinessInfo().getLocation().getSalePoint();
+      if (computeHaversine(clientLocation, vendorLocation) <= distanceLimit) {
+        nearbyVendors.add(vendor);
+      }    
+    }
+
+    return nearbyVendors;
+  }
+
+  /** Computes the distance between two geographical points using Haversine formula */
+  private float computeHaversine(GeoPt pointA, GeoPt pointB) {
```

Comment:
```suggestion
  private float computeDistance(GeoPt pointA, GeoPt pointB) {
```
Nit: I'd prefer if this was name computeDistance or something similar. The fact that the function uses the haversine formula is an implementation detail that the caller doesn't need to know about.

Reply:
Okay, makes sense to me!